src/nvim/eval.c

```diff
@@ -16753,6 +16855,83 @@ static void f_visualmode(typval_T *argvars, typval_T *rettv)
     curbuf->b_visual_mode_eval = NUL;
 }
 
+// watcheradd(dict, key, funcref) function
+static void f_watcheradd(typval_T *argvars, typval_T *rettv)
+{
+  if (check_restricted() || check_secure()) {
+    return;
+  }
+
+  if (argvars[0].v_type != VAR_DICT ||
+      argvars[1].v_type != VAR_STRING ||
+      (argvars[2].v_type != VAR_FUNC && argvars[2].v_type != VAR_STRING)) {
+    // Wrong argument types
+    EMSG(_(e_invarg));
+    return;
+  }
+
+  if (!STRLEN(argvars[1].vval.v_string)) {
+    EMSG(_("Can't use an empty string as the watched key"));
+    return;
+  }
+
+  ufunc_T *func = find_ufunc(argvars[2].vval.v_string);
+  if (!func) {
+    // Invalid function name. Error already reported by `find_ufunc`.
+    return;
+  }
+
+  func->uf_refcount++;
+  DictWatcher *watcher = xmalloc(sizeof(DictWatcher));
+  watcher->key_pattern = xstrdup((char *)argvars[1].vval.v_string);
```

Comment:
Since key length is already checked I would suggest to use

```C
const size_t key_len = STRLEN(argvars[1].vval.v_string);
if (key_len == 0) {
  EMSG(_(e_emptykey));
}
…
watcher->key_pattern = xmemdupz(argvars[1].vval.v_string, key_len)
```